torch/compiler/__init__.py

```diff
@@ -0,0 +1,319 @@
+
+def reset() -> None:
+    """
+    Clear all compile caches and restore the initial state.
+
+    This function clears all compile caches and resets various internal state variables to their initial values.
+
+    It performs the following operations:
+    1. Clears the input and output code caches.
+    2. Clears the original code map.
+    3. Clears the guard failures.
+    4. Clears the cache for the `ContinueExecutionCache` class.
+    5. If the most recent backend in the evaluation frame has a 'reset' method, it is called.
+    6. Resets the most recent backend in the evaluation frame to None.
+    7. Clears the compilation metrics.
+    8. Resets the frame count to its initial value.
+
+    Note:
+    - Call this function if you're compiling more than one model in a program or if you're changing the compile options
+
+    """
+
+    torch._dynamo.reset()
+
+def allow_in_graph(fn):
+    """
+    Customize which functions compilation will include in the generated graph.
+    Similar to `torch.fx.wrap()` it can be used as an escape hatch from graph breaks
+
+    Arguments:
+    - fn: A callable representing the function to be included in the graph.
+
+    Returns:
+    - If `fn` is a single callable, it adds the function to the list of allowed functions
+    in compilations internal storage and returns the function itself.
+    - If `fn` is a list or tuple of callables, it recursively applies the `allow_in_graph()`
+    function to each item in the list or tuple and returns a new list containing the
+    modified functions.
+
+    Note:
+    - The function assumes that `fn` is a callable. If it is not, an assertion error is raised.
+    """
+
+    return torch._dynamo.allow_in_graph(fn)
+
+
+def list_backends(exclude_tags=("debug", "experimental")) -> List[str]:
+    """
+    Return valid strings that can be passed to `torch.compile(..., backend="name")`.
+
+    Arguments:
+    - exclude_tags (optional): A tuple of strings representing tags to exclude.
+    Backends with any of the specified tags will not be included in the returned list.
+    By default, the tags "debug" and "experimental" are excluded.
+
+    Returns:
+    - A sorted list of backend names that can be passed to `torch.compile()`.
+
+    Note:
+    - The function internally performs a lazy import using the `_lazy_import()` function.
+    - The `exclude_tags` argument allows users to filter out backends based on their tags.
+    If the argument is not provided or is set to `None`, no tags will be excluded.
+
+    Example:
+    To retrieve a list of available backends excluding the tags "debug" and "experimental",
+    we can call the `list_backends()` function as follows:
+
+    ::
+        valid_backends = list_backends(exclude_tags=("debug", "experimental"))
+
+    """
+
+    return torch._dynamo.list_backends(exclude_tags)
+
+def explain(f, *args, **kwargs):
+    """
+    Run the function `f` with compilation and provide an explanation of the optimization process.
```

Comment:
cc @voznesenskym  to review

Reply:
Yo @voznesenskym do you explain is ready to be public? It's great I think the BC guarantee isn't too stringent since the only guarantee is returning a tuple so far

Reply:
Hmm good, lemme remove explain from here for now and I can work on a seperate PR making explain nice